core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -205,4 +238,24 @@ private[fs2] object Algebra {
     }
     fr.translate[Algebra[G,O,?]](algFtoG)
   }
+
+  /**
+    * Specific version of `Free.asHandler` that will output any CloseScope that may have
+    * been opened before the `rsn` was yielded
+    */
+  def asHandler[F[_], O](free: FreeC[Algebra[F,O,?],Unit], e: Throwable): FreeC[Algebra[F,O,?],Unit] = {
```

Comment:
This seems too adhoc to me, like we're working around a flaw somewhere else in the design.

Reply:
Well this is all realted to https://github.com/functional-streams-for-scala/fs2/issues/1022. Any shortcuting of propert CloseScope causes the (A) resource leakage (B) interuption to be not possible. May be there is more princcipled approach, but that may // likely will require to touch algebra of either FreeC or Algebra itself

Reply:
Re: #1022, `handleErrorWith` can be defined like I suggested if we want it to introduce a scope and hence, release resources acquired during that stream. Critically, there's no resource leak there at all, just a mismatched expectation as to scoping.

Reply:
As a thought experiment, what would happen if `++` on `Stream` introduced a new scope? Would we be able to keep the old definition of `flatMap` then?

Reply:
It is not unfortunatelly helping (the issue with handleErrorWith). I will bring you example of resource leak thats because of the mismatched scoping later. See the issue is that with interrupt we need to interrupt flatMaps that are defined in program AFTER the scope was defined, however stays within the `scope` of execution. i.e.  
```scala
Stream(1).flatMap { i => xxx}.interruptWhen(...).scope // that will work
Stream(1).interruptWhen(...).scope.flatMap { i => xxx} // thats the problem
```

Reply:
Re ++'s scope. That was one of the earliest attempts. If we introduce scope with every ++, performance sucks.

Reply:
Yes I understand performance will be unacceptable, but my point is that all this adhoc machinery is really just trying to make up for lack of scopes on each append in a use case specific way. I'm really not convinced we should be doing this.

Reply:
Not sure I have good answer to that. If we won't we couldn't have interruption done properly (not just ++, but also error propagation) and that is cutting fs2 as option to define more complex programs. If we introduce scope to ++ we will sacrifice perfromance in place where we don't want to.

Reply:
I'd like to get some thoughts from @pchiusano -- maybe he has some ideas.

Reply:
Ok, sure.